### PR TITLE
feat(Table): UXD-1314 adding support for colspan and complex headers

### PR DESCRIPTION
**Pull request by nahumzs**
### Purpose 🚀
Add support for colspan to the Table component

Screen Shot 2021-07-12 at 11 14 37 AM

Screen Shot 2021-07-12 at 11 14 57 AM

### Notes ✏️
tests are on screener

### Updates 📦
If you have changed a component's source code (not stories, specs, or docs), before merging your branch run `yarn changeset`. This will prompt you to:
- indicate if changes were patch/minor/major for each modified package
- enter a release message

### Storybook 📕
http://storybooks.highbond-s3.com/paprika/your-branch-name

### Screenshots 📸
_optional but highly recommended_

### References 🔗
_relevant Jira ticket / GitHub issues_

### CODE REVIEW

File: packages/Table/src/Table.js [hunk 1]
```diff
@@ -24,41 +26,55 @@ const Table = React.forwardRef((props, ref) => {
...
-                <sc.TH sticky={sticky} borderType={borderType} key={columnIndex} {...moreColumnProps}>
-                  {header}
-                </sc.TH>
-              );
-
-            throw new Error("Header should be either of type string or function");
-          })}
-        </tr>
-      </sc.Thead>
+      {hasAtLeastOneHeader ? (
```
Comment (mikrotron) [inline packages/Table/src/Table.js:50]:
```suggestion
      {hasAtLeastOneHeader || extractedHeaders ? (
```

File: packages/Table/src/Table.styles.js [hunk 2]
```diff
@@ -3,21 +3,23 @@ import tokens from "@paprika/tokens";
...
 
-  & .is-highlighted-focus {
-    outline: 2px solid blue;
-  }
-  & .is-highlighted-idle {
-    outline: 2px solid transparent;
-  }
-`;
+    & .is-highlighted-focus {
+      outline: 2px solid blue;
```
Comment (mikrotron) [inline packages/Table/src/Table.styles.js:16]: A little out of scope, but maybe a quick token swap in would be a quick win?

Probably `tokens.highlight.active.noBorder.borderColor` is the best fit, or else just `tokens.color.blue`.

File: packages/Table/src/index.js [hunk 3]
```diff
@@ -1 +1,2 @@
+export { TH } from "./Table.styles";
```
Comment (mikrotron) [inline packages/Table/src/index.js:1]: Do we export a styled component in the index of any of our other components?

Would there be a more obvious or scalable API to expose them?

Like, as `<Table.SC.TH>`?  I'm not crazy about that... but actually, maybe just exporting as `<Table.TH>` would be fine – the fact that it's just a _styled_ component doesn't really matter to the consumer.

File: packages/Table/stories/Table.colspans.stories.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        cell={({ row }) => {
+          return `${row.lastName}`;
+        }}
+      />
+      <Table.ColumnDefinition header="ice cream" cell="iceCream" />
+    </Table>
+  );
+};
+
+export const WithoutHeaderPropAndTableHeader = () => {
```
Comment (mikrotron) [inline packages/Table/stories/Table.colspans.stories.js:42]: This one doesn't look right. Was that intentional?

Screen Shot 2021-07-12 at 11 43 18 AM

File: packages/Table/stories/Table.colspans.stories.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+];
+
+export const WithColSpan = () => {
+  return (
+    <Table a11yText="" data={data}>
+      <Table.Headers>
+        <tr>
+          <TH colSpan="3">List of ice creams</TH>
+        </tr>
+      </Table.Headers>
```
Comment (mikrotron) [inline packages/Table/stories/Table.colspans.stories.js:30]: Can we include this detail that the `children` of `<Table.Headers>` should include a `<tr>` in the Docs MDX page?  Probably just a sentence or two at the bottom of that page under "Custom Headers" would be enough.

**Comment by allison-c** [comment]
A general question, do we really need to support `colspan` with these kinds of table APIs(ColumnsDefinition)? asking because there might be a case we need a `colspan` on a `td`, then how can we tell the `colSpan` prop on the `ColumnDefinition` should apply to the `th` or `td` ? if it's `td`, how can we tell which `td` should has `colspan`...etc
I think we might need a more flexible render method if we need to support `colspan` on `td` or even `rowspan` later on

File: packages/Table/src/Table.js (hunk 1 above)
Comment (nahumzs) [reply packages/Table/src/Table.js:51]: if fixed this already differently

File: packages/Table/src/Table.styles.js (hunk 2 above)
Comment (nahumzs) [reply packages/Table/src/Table.styles.js:16]: `tokens.highlight.active.noBorder.borderColor` wao, really complex name for a token, wondering the cognitive load to remember this name :D will replace it.

File: packages/Table/src/index.js (hunk 3 above)
Comment (nahumzs) [reply packages/Table/src/index.js:1]: it has to be used, in order to match the border types and even if it's sticky or not. This only affect the edge case of complex header which I think is sane to do it in this way

File: packages/Table/stories/Table.colspans.stories.js (hunk 4 above)
Comment (nahumzs) [reply packages/Table/stories/Table.colspans.stories.js:42]: you were caught in the middle of an update this has been remove, try to refresh the header :D sorry.

**Comment by nahumzs** [comment]
> A general question, do we really need to support `colspan` with these kinds of table APIs(ColumnsDefinition)? asking because there might be a case we need a `colspan` on a `td`, then how can we tell the `colSpan` prop on the `ColumnDefinition` should apply to the `th` or `td` ? if it's `td`, how can we tell which `td` should has `colspan`...etc
> I think we might need a more flexible render method(e.g. a render method devs can use and skip ColumnDefinition) if we need to support `colspan` on `td` or even `rowspan` later on

I thought about this as well and I realized that we really don't need it since the ColumnDefinitions can be used to do it without a need of colspan

```js
<Table.ColumnDefinion cell={({ row }) => {
	return <div>{row.columData1} {row.columData2}</div>
}} />
```

The previous combine two columns in one.

File: packages/Table/src/Table.js [hunk 6]
```diff
@@ -24,41 +27,57 @@ const Table = React.forwardRef((props, ref) => {
...
+                  </sc.TH>
+                );
+
+              throw new Error("Header should be either of type string or function");
+            })}
+          </tr>
+        </sc.Thead>
+      ) : extractedHeaders ? (
+        <sc.Thead>{extractedHeaders}</sc.Thead>
+      ) : null}
```
Comment (AndreyChernykh) [inline packages/Table/src/Table.js:80]: Conditions like that `? : ? :` is hard to read and understand. 
I'd prefer it that case just move this block into a `renderHeader()` function and use a regular `if else` statements instead

Comment (AndreyChernykh) [reply packages/Table/src/Table.js:80]: Also, there is enough logic (including `hasAtLeastOneHeader`) to consider extracting all that into a separate internal component, I think

**Comment by allison-c** [comment]
> > A general question, do we really need to support `colspan` with these kinds of table APIs(ColumnsDefinition)? asking because there might be a case we need a `colspan` on a `td`, then how can we tell the `colSpan` prop on the `ColumnDefinition` should apply to the `th` or `td` ? if it's `td`, how can we tell which `td` should has `colspan`...etc
> > I think we might need a more flexible render method(e.g. a render method devs can use and skip ColumnDefinition) if we need to support `colspan` on `td` or even `rowspan` later on
> 
> I thought about this as well and I realized that we really don't need it since the ColumnDefinitions can be used to do it without a need of colspan
> 
> ```js
> <Table.ColumnDefinion cell={({ row }) => {
> 	return <div>{row.columData1} {row.columData2}</div>
> }} />
> ```
> 
> The previous combine two columns in one.

So if I do that ☝️ , the html will look like this?
```jsx
<table>
  <tr>
    <td><div>data1 data2</div></td> // or <td colspan="2"><div>data1 data2</div></td>
    <td></td>
  <tr>
</table>
```

File: packages/Table/stories/Table.colspans.stories.js (hunk 5 above)
Comment (nahumzs) [reply packages/Table/stories/Table.colspans.stories.js:30]: added documentation, can you take a look :)

**Comment by nahumzs** [comment]
```jsx
<table>
  <tr>
    <td><div>data1 data2</div></td>
    <td></td>
  <tr>
</table>
```

would not be concept of `colspan` what do you think would be the downside for that, thinking as well that this is not a common case?

**Comment by allison-c** [comment]
> ```js
> <table>
>   <tr>
>     <td><div>data1 data2</div></td>
>     <td></td>
>   <tr>
> </table>
> ```
> 
> would not be concept of `colspan` what do you think would be the downside for that, thinking as well that this is not a common case?

Probably not a common case, it just doesn't look semantically right for a long term, I'll need to think about it 😅

**Review by mikrotron — APPROVED** [review]
llgtm 72

**Pull request merged**